src/Toml/PrefixTree.hs

```diff
@@ -0,0 +1,101 @@
+module Toml.PrefixTree
+       ( PrefixTree (..)
+       , single
+       , insert
+       , lookup
+       , delete
+       ) where
+
+import Prelude hiding (lookup)
+
+import Data.Hashable (Hashable)
+import Data.HashMap.Strict (HashMap)
+import Data.List.NonEmpty (NonEmpty)
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Text (Text)
+
+import qualified Data.HashMap.Strict as HashMap
+
+-- | Represents the key piece of some layer.
+newtype Piece = Piece { unPiece :: Text }
+    deriving (Show, Eq, Hashable)
+
+-- | Full representation of Key contains pieces of every layer.
+newtype Key = Key { unKey :: NonEmpty Piece }
+    deriving (Show, Eq)
+
+-- | Type synonym for 'Key'.
+type Prefix = Key
+
+-- | Map of layer names and corresponding 'PrefixTree's.
+type PrefixMap a = HashMap Piece (PrefixTree a)
+
+-- | Data structure to represent @toml@.
+data PrefixTree a
+    = Leaf Key a
+    | Branch { bcommonPref :: Prefix         -- ^ greatest common prefix
+             , bVal        :: (Maybe a)      -- ^ value by key = prefix
+             , bPrefixMap  :: (PrefixMap a)  -- ^ suffixes of prefix
+             }
+
+data KeysDiff
+      -- | Keys are equal
+    = Equal
+      -- | Keys don't have any common part.
+    | NoPrefix
+      -- | The first key is the prefix for the second one.
+    | FstIsPref { diff :: Key}
+      -- | The second key is the prefix for the first one.
+    | SndIsPref { diff :: Key}
+      -- | Key have same prefix.
+    | Diff { pref    :: Key
+           , diffFst :: Key
+           , diffSnd :: Key
+           }
+    deriving (Show, Eq)
+
+toKey :: Piece -> [Piece] -> Key
+toKey x xs = Key $ x :| xs
+
+keysDiff :: Key -> Key -> KeysDiff
+keysDiff (Key (x :| xs)) (Key (y :| ys)) =
+    if x == y
+    then listSame xs ys []
+    else NoPrefix
+  where
+    listSame [] []     _ = Equal
+    listSame [] (s:ss) _ = FstIsPref $ toKey s ss
+    listSame (f:fs) [] _ = SndIsPref $ toKey f fs
+    listSame (f:fs) (s:ss) pr =
+        if f == s
+        then listSame fs ss (pr ++ [f])
```

Comment:
Appending element to the end of this list if very slow operation, unfortunately... Especially when such appends accumulates. Calling `reverse` function only once is a better solution but still very slow. Since this function will be called on every `lookup` and even multiple times per lookup, it's better to optimize it. Though, it's not immediately obvious to me how to do, this requires some techical details... Moreover, I don't think this is really required. In most cases `Prefix` will be just singleton map. So this `++` will be called really rare. 

Otherwise it  looks completely correct :slightly_smiling_face:  So let's leave it as it is!